did_peer/src/numalgos/numalgo2/resolve/helpers.rs

```diff
@@ -0,0 +1,413 @@
+    }
+
+    Ok(did_doc_builder)
+}
+
+fn process_service_element(
+    element: &str,
+    mut did_doc_builder: DidDocumentBuilder<ExtraFieldsSov>,
+    service_index: &mut usize,
+) -> Result<DidDocumentBuilder<ExtraFieldsSov>, DidPeerError> {
+    let decoded = STANDARD_NO_PAD.decode(element)?;
+    let service: OneOrList<ServiceAbbreviated> = serde_json::from_slice(&decoded)?;
+
+    match service {
+        OneOrList::One(service) => {
+            did_doc_builder = did_doc_builder.add_service(deabbreviate_service(service, *service_index)?);
+            *service_index += 1;
+        }
+        OneOrList::List(services) => {
+            for service in services.into_iter() {
+                did_doc_builder = did_doc_builder.add_service(deabbreviate_service(service, *service_index)?);
+                *service_index += 1;
+            }
+        }
+    }
+
+    Ok(did_doc_builder)
+}
+
+fn process_key_element(
+    element: &str,
+    mut did_doc_builder: DidDocumentBuilder<ExtraFieldsSov>,
+    did: &Did,
+    public_key_encoding: &PublicKeyEncoding,
+    purpose: ElementPurpose,
+) -> Result<DidDocumentBuilder<ExtraFieldsSov>, DidPeerError> {
+    let key = Key::from_fingerprint(&element)?;
+    let vms = get_verification_methods_by_key(&key, did, public_key_encoding)?;
+
+    for vm in vms.into_iter() {
+        match purpose {
+            ElementPurpose::Assertion => {
+                did_doc_builder = did_doc_builder.add_assertion_method(vm);
+            }
+            ElementPurpose::Encryption => {
+                did_doc_builder = did_doc_builder.add_key_agreement(vm);
+            }
+            ElementPurpose::Verification => {
+                did_doc_builder = did_doc_builder.add_verification_method(vm);
+            }
+            ElementPurpose::CapabilityInvocation => did_doc_builder = did_doc_builder.add_capability_invocation(vm),
+            ElementPurpose::CapabilityDelegation => did_doc_builder = did_doc_builder.add_capability_delegation(vm),
+            _ => return Err(DidPeerError::UnsupportedPurpose(purpose.into())),
+        }
+    }
+
+    Ok(did_doc_builder)
+}
+
+fn deabbreviate_service(service: ServiceAbbreviated, index: usize) -> Result<Service<ExtraFieldsSov>, DidPeerError> {
+    let service_type = match service.service_type() {
+        "dm" => "DIDCommMessaging".to_string(),
+        t @ _ => t.to_string(),
+    };
+
+    let id = format!("#{}-{}", service_type.to_lowercase(), index).parse()?;
+
+    if service.routing_keys().is_empty() {
+        build_service_aip1(service, id, service_type)
+    } else {
+        build_service_didcommv2(service, id, service_type)
+    }
+}
+
+fn build_service_aip1(
+    service: ServiceAbbreviated,
+    id: Uri,
+    service_type: String,
+) -> Result<Service<ExtraFieldsSov>, DidPeerError> {
+    Ok(Service::<ExtraFieldsSov>::builder(
+        id,
+        service.service_endpoint().parse()?,
```

Comment:
why not to return `Url` from `service_endpoint()`? Then we wouldn't have re-parse what we already are sure is valid url